xds/src/main/java/io/grpc/xds/LookasideLb.java

```diff
@@ -170,48 +181,137 @@ public void handleNameResolutionError(Status error) {}
+                          GrpcUtil.STOPWATCH_SUPPLIER, node);
+                    }
+                  });
+                } else {
+                  lrsClient = cachedLrsClient;
+                  xdsClientRef = cachedXdsClientRef;
+                }
+              } else if (cachedXdsClientRef == null) {
+                // This is to handle EDS-only with bootstrap usecase.
+                final BootstrapInfo bootstrapInfo;
+                try {
+                  bootstrapInfo = bootstrapper.readBootstrap();
+                } catch (Exception e) {
+                  helper.updateBalancingState(
+                      ConnectivityState.TRANSIENT_FAILURE,
+                      new ErrorPicker(Status.UNAVAILABLE.withCause(e)));
+                  return;
+                }
+                final ManagedChannel channel = initLbChannel(
+                    helper, bootstrapInfo.getServerUri(),
+                    bootstrapInfo.getChannelCredentials());
                 // TODO(zdapeng): Use XdsClient to do Lrs directly.
                 lrsClient = loadReportClientFactory.createLoadReportClient(
-                    channel, helper, new ExponentialBackoffPolicy.Provider(),
-                    loadStatsStore);
-                final LoadReportCallback lrsCallback =
-                    new LoadReportCallback() {
-                      @Override
-                      public void onReportResponse(long reportIntervalNano) {
-                        localityStore.updateOobMetricsReportInterval(reportIntervalNano);
-                      }
-                    };
-
-                EndpointWatcher endpointWatcher =
-                    new EndpointWatcherImpl(lrsClient, lrsCallback, localityStore);
-                xdsClient.watchEndpointData(node.getCluster(), endpointWatcher);
+                    channel, helper, new ExponentialBackoffPolicy.Provider(), loadStatsStore);
+                xdsClientRef = new RefCountedXdsClientObjectPool(new XdsClientFactory() {
+                  @Override
+                  XdsClient createXdsClient() {
+                    return new XdsComms2(
+                        channel, helper, new ExponentialBackoffPolicy.Provider(),
+                        GrpcUtil.STOPWATCH_SUPPLIER, bootstrapInfo.getNode());
+                  }
+                });
+              } else {
+                // This is the XdsResolver or non EDS-only usecase.
+
+                // FIXME(zdapeng): Use XdsClient to do Lrs directly.
+                // No-op for now.
+                lrsClient = new LoadReportClient() {
+                  @Override
+                  public void startLoadReporting(LoadReportCallback callback) {}
+
+                  @Override
+                  public void stopLoadReporting() {}
+                };
+                xdsClientRef = cachedXdsClientRef;
               }
+
+              cachedLrsClient = lrsClient;
+              cachedXdsClientRef = xdsClientRef;
```

Comment:
The back and forth assignments between `cachedXdsClientRef` and `xdsClientRef` (e.g. lines 265 and 269) are confusing since this is Object pool and the getObject and returnObject have to match.

Reply:
I added more explanation comments. Basically once `xdsClientRef` is assigned (by either creating a new one or loading the existing cachedXdsClientRef), it will be cached for later use.

Reply:
I reorganized the code a little bit. Hopefully it's easier to read than before.